### PR TITLE
Define Interactor arguments explicitly to improve testability

### PR DESCRIPTION
**Pull request by dostu**
**Problem**
Currently stubbing interactor calls is a risky activity. If you add an additional attribute to the `initialize_with` RSpec won't notice anything wrong from the call site if the interactor call is stubbed. I've seen this break production functionality multiple times, including for me. Example: https://github.com/vinted/core/pull/75015
 
**Solution**
Define methods in a less dynamic way so RSpec can see that the interactor call is stubbed with wrong argument count. 

Example: 
Screenshot 2023-05-23 at 11 49 58

WDYT? I could work on improving the PR code quality if it's something we could use.

@vinted/platform-backend

### CODE REVIEW

File: spec/interactor/initializer_spec.rb [hunk 1]
```diff
@@ -68,19 +68,19 @@ def run
     end
 
     describe '.for' do
-      subject { interactor.for(params) }
+      subject { interactor.for(**params) }
```
Comment (tomasv) [inline spec/interactor/initializer_spec.rb]: Does this mean that this is a breaking change? Why did you need to add `**`?

**Review by tomasv — APPROVED** [review]
Don't forget to bump version too and update Gemfile.lock.

File: spec/interactor/initializer_spec.rb (hunk 1 above)
Comment (dostu) [reply spec/interactor/initializer_spec.rb]: Reverted the keyword params class methods because RSpec won't notice changes when stubbing keyword arguments anyway.

**Comment by dostu** [comment]
It turns out this is a breaking change because there are some interactor classes in core which override initializer to get optional params:

```
include Interactor::Initializer

initialize_with :data, :sync

def initialize(data, sync = false)
  @data = data
  @sync = sync
end		
```

I've updated all the cases with custom class methods here: https://github.com/vinted/core/pull/75066/files. @tomasv do you think it's OK to remove this workaround for optional params?

**Comment by tomasv** [comment]
OK, I guess, too bad we can't support default arguments now.

**Comment by andriusm** [comment]
Interesting idea, but I don't know if this will be very useful. As I understand, it would not prevent cases when the number of arguments matches, but the argument itself is of incorrect type (recent case https://github.com/vinted/core/pull/75575). I believe the issue is that the very thing that's being tested is also being mocked.

If you decide to merge this, at least increase the minor version, not just the patch version.

**Comment by tomasv** [comment]
> As I understand, it would not prevent cases when the number of arguments matches, but the argument itself is of incorrect type.

That's a different class of errors, I don't think we can catch those without extra type annotations like Sorbet.

**Comment by andriusm** [comment]
> That's a different class of errors, I don't think we can catch those without extra type annotations like Sorbet.

We can't and that's my point - merging this change may provide a false sense of security to developers.

**Comment by tomasv** [comment]
> We can't and that's my point - merging this change may provide a false sense of security to developers.

I don't understand your point. We should not merge this beause it prevents only some errors but not all and therefore we should not do it at all? A typical developer won't even see this change, only his specs will catch more errors. That's a net good.

**Comment by andriusm** [comment]
If you consider it a net good then fine.

**Comment by oleg-vinted** [comment]
Can you elaborate on why the same approach does not work with keyword params?

**Comment by dostu** [comment]
> Can you elaborate on why the same approach does not work with keyword params?

Actually it would work, specs would catch missing keywords even with stubbing:

Screenshot 2023-06-12 at 12 30 59

It would be a breaking change though, because currently we accept a hash: https://github.com/vinted/interactor-initializer/blob/master/lib/interactor/initializer/helper.rb#L45

I'll check how much work it would be to adjust core code to this change.

**Comment by GediminasZukas** [comment]
At shipping we also recently faced the production env [incident](https://github.com/vinted/svc-shipping/pull/10374) due this issue.
